src/state_space/cache.rs

```diff
@@ -0,0 +1,89 @@
+use std::collections::HashMap;
+
+use crate::amm::{AutomatedMarketMaker, AMM};
+
+use super::StateChange;
+use arraydeque::{ArrayDeque, CapacityError};
+
+#[derive(Debug)]
+
+pub struct StateChangeCache {
+    oldest_block: u64,
+    cache: ArrayDeque<StateChange, 150>,
+}
+
+impl StateChangeCache {
+    pub fn new() -> Self {
+        StateChangeCache {
+            oldest_block: 0,
+            cache: ArrayDeque::new(),
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.cache.is_empty()
+    }
+
+    pub fn add_state_change_to_cache(
+        &mut self,
+        state_change: StateChange,
+    ) -> Result<(), CapacityError<StateChange>> {
+        let cache = &mut self.cache;
+
+        if cache.is_full() {
+            cache.pop_back();
+            self.oldest_block = cache
+                .back()
+                .expect("Could not get last item in cache")
```

Comment:
We can actually `unwrap()` here since we first check if the cache is full. Since the cache has a const size of 150 items, we know there is at least item in the dequeue.